### PR TITLE
Random managed buffers

### PR DESCRIPTION
**Pull request by andrei-diaconescu**

### CODE REVIEW

File: arwen/elrondapi/manBufOps.go [hunk 1]
```diff
@@ -393,3 +399,28 @@ func v1_4_mBufferFinish(context unsafe.Pointer, mBufferHandle int32) int32 {
...
 	return 0
 }
+
+//export v1_4_mBufferNewRandom
+func v1_4_mBufferNewRandom(context unsafe.Pointer, length int32) int32 {
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	if length < 1 {
```
Comment (sasurobert) [inline arwen/elrondapi/manBufOps.go:409]: _ = arwen.WithFault

File: arwen/elrondapi/manBufOps.go [hunk 2]
```diff
@@ -393,3 +399,28 @@ func v1_4_mBufferFinish(context unsafe.Pointer, mBufferHandle int32) int32 {
...
+		arwen.WithFault(arwen.ErrLengthOfBufferNotCorrect, context, runtime.ManagedBufferAPIErrorShouldFailExecution())
+		return -1
+	}
+
+	baseGasToUse := metering.GasSchedule().ManagedBufferAPICost.MBufferNewFromBytes
+	lengthDependentGasToUse := math.MulUint64(metering.GasSchedule().BaseOperationCost.DataCopyPerByte, uint64(length))
+	metering.UseGas(baseGasToUse + lengthDependentGasToUse)
```
Comment (sasurobert) [inline arwen/elrondapi/manBufOps.go:416]: you have to use math.AddUint64

File: arwen/elrondapi/manBufOps.go [hunk 3]
```diff
@@ -393,3 +399,28 @@ func v1_4_mBufferFinish(context unsafe.Pointer, mBufferHandle int32) int32 {
...
+	managedType := arwen.GetManagedTypesContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	if length < 1 {
+		arwen.WithFault(arwen.ErrLengthOfBufferNotCorrect, context, runtime.ManagedBufferAPIErrorShouldFailExecution())
+		return -1
+	}
+
+	baseGasToUse := metering.GasSchedule().ManagedBufferAPICost.MBufferNewFromBytes
```
Comment (sasurobert) [inline arwen/elrondapi/manBufOps.go:414]: use MBufferNewRandom instead

File: arwen/elrondapi/manBufOps.go [hunk 4]
```diff
@@ -393,3 +399,28 @@ func v1_4_mBufferFinish(context unsafe.Pointer, mBufferHandle int32) int32 {
...
+	lengthDependentGasToUse := math.MulUint64(metering.GasSchedule().BaseOperationCost.DataCopyPerByte, uint64(length))
+	metering.UseGas(baseGasToUse + lengthDependentGasToUse)
+
+	randomizer := managedType.GetRandReader()
```
Comment (sasurobert) [inline arwen/elrondapi/manBufOps.go:418]: one more change for getRandReader - make it use previous block randomness, current block randomness and tx hash

File: arwen/contexts/managedType.go [hunk 5]
```diff
@@ -60,7 +60,7 @@ func NewManagedTypesContext(host arwen.VMHost) (*managedTypesContext, error) {
 }
 
 func (context *managedTypesContext) initRandomizer() {
-	randomizer := math.NewSeedRandReader(append(context.host.Blockchain().CurrentRandomSeed(), context.host.Runtime().GetCurrentTxHash()...))
+	randomizer := math.NewSeedRandReader(append(append(context.host.Blockchain().CurrentRandomSeed(), context.host.Runtime().GetCurrentTxHash()...), context.host.Blockchain().LastRandomSeed()...))
```
Comment (sasurobert) [inline arwen/contexts/managedType.go:63]: beutify this please.

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

**Review by camilbancioiu — DISMISSED** [review]

File: test/contracts/managed-buffers/managed-buffers.c [hunk 6]
```diff
@@ -86,10 +86,144 @@ void mBufferMethod() {
...
     
     finishResult(ok);
 }
 
+void mBufferNewTest() {
+    int reps;
+    int64getArgument(reps);
+    for (int i = 0; i < reps; i++)
+    {
+        int handle = mBufferNew();
```
[the remark this answers is not in the document]
Comment (andrei-marinica) [reply test/contracts/managed-buffers/managed-buffers.c:102]: I mentioned above but not for all of them, I think the best way would be to return the resulting buffers themselves, using `mBufferFinish`

**Review by andrei-marinica — DISMISSED** [review]

**Review by sasurobert — DISMISSED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by andrei-marinica — APPROVED** [review]

**Pull request merged**